server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -868,4 +876,76 @@ public void onFailure(Exception e) {
+     * be visible in the logs.
+     */
+    // visible for testing
+    class ConsistencyChecker {
+
+        // dedicated logger so we can separate these logs from other SnapshotShardsService DEBUG logs
+        private static final Logger CONSISTENCY_CHECKER_LOGGER = LogManager.getLogger(ConsistencyChecker.class);
+
+        private record CheckTask(Snapshot snapshot, ShardId shardId) {}
+
+        private final AtomicInteger queuedTaskCount = new AtomicInteger(0); // atomic to ensure max one thread processing the queue
+        private final Queue<CheckTask> queue = new ConcurrentLinkedQueue<>();
+
+        void ensureShardComplete(Snapshot snapshot, ShardId shardId) {
+            if (CONSISTENCY_CHECKER_LOGGER.isDebugEnabled() == false) {
+                return;
+            }
+
+            if (queuedTaskCount.get() > 1000) {
+                // racy check, we only need an approximate limit
+                return;
+            }
+
+            queue.add(new CheckTask(snapshot, shardId));
+            if (queuedTaskCount.getAndIncrement() == 0) {
+                threadPool.generic().execute(this::runCheck);
+            } // else a runCheck is already running somewhere and will pick up the task we just added
+        }
+
+        private void runCheck() {
+            while (true) {
+                final var taskCount = queuedTaskCount.get();
+                final var shardsBySnapshot = new HashMap<Snapshot, Set<ShardId>>();
+                for (int i = 0; i < taskCount; i++) {
+                    final var task = queue.poll();
+                    assert task != null;
+                    shardsBySnapshot.computeIfAbsent(task.snapshot(), ignored -> new HashSet<>()).add(task.shardId());
+                }
+                final var snapshotsInProgress = SnapshotsInProgress.get(clusterService.state());
+                for (final var shardsBySnapshotEntry : shardsBySnapshot.entrySet()) {
+                    final var snapshot = shardsBySnapshotEntry.getKey();
+                    final var entry = snapshotsInProgress.snapshot(snapshot);
+                    if (entry != null) {
+                        for (final var shardId : shardsBySnapshotEntry.getValue()) {
+                            final var shardStatus = entry.shards().get(shardId);
+                            if (shardStatus == null) {
+                                CONSISTENCY_CHECKER_LOGGER.debug("shard [{}] in snapshot [{}] unexpectedly not found", shardId, snapshot);
+                            } else if (shardStatus.state().completed() == false) {
+                                CONSISTENCY_CHECKER_LOGGER.debug(
+                                    "shard [{}] in snapshot [{}] unexpectedly still in state [{}] after notifying master",
+                                    shardId,
+                                    snapshot,
+                                    shardStatus
+                                );
```

Comment:
Hmm. So as soon as a shard snapshot status update action completes on the data node, the data node will immediately check the current cluster state. Do you know if it is expected that the master node successfully publishes the new cluster state before replying to the data node? 

Cluster state update driven verification could technically fail as well, if no cluster state update is received after the data node checks and doesn't see the change. I wonder if throwing these failed checks onto another queue for a periodic checker, which warns, would make this solution complete (no races): the periodic warnings would just continue endlessly until a data node restart or master change.

Reply:
Yes, the master node publishes the state and waits for a reasonable length of time (`cluster.publish.timeout`, 30s by default) for all the other nodes to confirm they have applied the state before it completes any waiting listeners.

Reply:
> Yes, the master node publishes the state and waits for a reasonable length of time (cluster.publish.timeout, 30s by default) for all the other nodes to confirm they have applied the state before it completes any waiting listeners.

Great, that greatly reduces the chance of racing I was imagining.

> But these checks are happening exactly after the cluster state update that should have updated the state. We're checking that the state update we asked for has happened. We don't need to do this periodically, once is enough, at least for debugging purposes.

Agreed, this should be sufficient. The edge case explanation in the class comment is helpful, too 👍